src/Microsoft.IdentityModel.Tokens/Cache/MaximumSizeCache.cs

```diff
@@ -0,0 +1,272 @@
+        public virtual void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
+        }
+
+        /// <inheritdoc/>
+        public virtual bool SetValue(TKey key, TValue value, DateTime expirationTime)
+        {
+            ValidateValues(key, value, expirationTime);
+            if (NeedsCompaction && Interlocked.CompareExchange(ref _compactionState, CompactionStateInProgress, CompactionStateNotRunning) == CompactionStateNotRunning)
+                ThreadPool.QueueUserWorkItem(CompactCache);
+
+            _map[key] = new CacheItem<TKey, TValue>(key, value, expirationTime);
+
+            return true;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            value = default;
+            var found = TryGetValueInternal(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val.Value;
+
+            return found;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool TryRemove(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            value = default;
+
+            var found = TryRemoveInternal(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val.Value;
+
+            return found;
+        }
+
+        #endregion
+
+
+        #region protected
+
+        /// <summary>
+        /// The internal version of TryGetValue that gets the LRUCacheItem from the hash.
+        /// </summary>
+        /// <param name="key">The key of the value to get.</param>
+        /// <param name="value">The object that has the specified key, or the default value of the type if the not found.</param>
+        /// <returns>true if the key was found, otherwise, false.</returns>
+        protected virtual bool TryGetValueInternal(TKey key, out CacheItem<TKey, TValue> value)
+        {
+            value = default;
+
+            var found = _map.TryGetValue(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val;
+
+            return found;
+        }
+
+        /// <summary>
+        /// Attempts to remove and return the value that has the specified key from the cache.
+        /// </summary>
+        /// <param name="key">The key of the element to remove and return.</param>
+        /// <param name="value">The object removed from the cache, or the default value of the TValue type if key does not exist.</param>
+        /// <returns>true if the object was removed successfully; otherwise, false.</returns>
+        protected virtual bool TryRemoveInternal(TKey key, out CacheItem<TKey, TValue> value)
+        {
+            value = default;
+
+            var found = _map.TryRemove(key, out CacheItem<TKey, TValue> val);
```

Comment:
as above could this be:
could you write this as:
```cs
return _map.TryRemoveValue(key, out CacheItem<TKey, TValue> value);
```

Reply:
It has been changed to `TryRemoveCachItem()` that wraps the `_map.TryRemove(key, out CacheItem<TKey, TValue> val)` call and shared by both classes.